### PR TITLE
ConfigManager to have ZDB backened

### LINKED ISSUE #25: ConfigManager to have configurable backend (redundant)
#### Detailed description:
ConfigManager to use FileSystem by default as before, but can also be configured to instead use 0-DB as a backend.

#### Requirements:
 - Encrypted keys should still be encrypted.
 - One config repo per system
 - Sandboxed config repos should still work as before (using their own keys)

#### Questions to resolve:
 - Are the key.seed and key.priv on the filesystem or in the DB?
 - Are jumpscale configuration and state files (where we keep previously installed modules and jumpscale plugins and config repo) on FS or DB?
 - is the SSH key used for encryption on FS or DB

### PR DESCRIPTION
**Pull request by xmonader**
#### What this PR resolves:
Fixes #25  (allows isolated configurations using zdb namespaces)

*Please review carefully*

### CODE REVIEW

File: Jumpscale/data/nacl/NACL.py [hunk 1]
```diff
@@ -93,15 +93,19 @@ def getagent(name):
             raise RuntimeError("Could not find agent for key with name:%s" % name)
 
         if self._agent is None:
-            if not j.clients.sshkey.exists(self.sshkeyname):
+            # Should we let keypath set in both DbConfigManager and FileConfigManager? What would be the behavior?
```
Comment (zaibon) [inline Jumpscale/data/nacl/NACL.py]: this doesn't look like it's doing the same as before, shouldn't it be:
```python
if self._agent is None:
    # Should we let keypath set in both DbConfigManager and FileConfigManager? What would be the behavior?
    if not j.clients.sshkey.exists(self.sshkeyname):
        keypath = j.tools.configmanager.keypath
        if not j.sal.fs.exists(keypath):
            keypath = "%s/.ssh/%s" % (j.dirs.HOMEDIR, self.sshkeyname)
        
        if j.sal.fs.exists(keypath):
            j.clients.sshkey.key_load(keypath)
        else:
            # if sshkeyname from state is not reachable delete it and re-init config manager
            j.core.state.configSetInDict("myconfig", "sshkeyname", "")
            j.tools.configmanager.init()
    
    self._agent = getagent(self.sshkeyname)

return self._agent
```

File: Jumpscale/tools/configmanager/DbConfig.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+    @property
+    def template(self):
+        if self._template is None or self._template == '':
+            obj = eval(self.location)
+            if hasattr(obj,"_child_class"):
+                obj._child_class
+                myvars={}
+                try:
+                    exec("from %s import TEMPLATE;template=TEMPLATE"%obj._child_class.__module__,myvars)
```
Comment (zaibon) [inline Jumpscale/tools/configmanager/DbConfig.py]: can't we avoid this exec ? this is vary bad practice

File: Jumpscale/tools/configmanager/DbConfig.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+    def save(self):
+        hsetmany(self._zdbsimplecl, self.hsetname, **self._data)
+
+    
+    def delete(self):
+        hdel(self._zdbsimplecl, self.hsetname)
+
+    @property
+    def template(self):
```
Comment (zaibon) [inline Jumpscale/tools/configmanager/DbConfig.py]: globablly I don't understand this method at all. It is quite complex, not possible to simplify or split in smaller chunks ?

File: Jumpscale/tools/configmanager/DbConfigManager.py [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+
+        configs_keys = iselect_all(self._zdbsimplecl, hsetname)
+        if force or j.tools.console.askYesNo("No location specified, Are you sure you want to delete all configs?", default=True):
+            for k in configs_keys:
+                hdel(self._zdbsimplecl, k)
+
+    @property
+    def path(self):
+        # rootconfig
+        from pathlib import Path
```
Comment (zaibon) [inline Jumpscale/tools/configmanager/DbConfigManager.py]: why do you need a new lib for this ?
And why an import in a property ? import belong to the beginning of a module

File: Jumpscale/tools/configmanager/README.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+secrets = ""
+```
+- set the backend type to `db` to use 0-db backend (it's `file` by default.)
+- You set the backend url (e.g localhost:9000)
+- secrets/adminsecrets from 0-db terminology
+- while using the 0-db backend you must call `j.tools.configmanager.set_namespace` first
+
+
+
+## Example
```
Comment (zaibon) [inline Jumpscale/tools/configmanager/README.md]: I don't understand what is going on in the example, can you put some comments in there ?

File: Jumpscale/data/nacl/NACL.py (hunk 1 above)
Comment (xmonader) [reply Jumpscale/data/nacl/NACL.py]: updated that

File: Jumpscale/tools/configmanager/DbConfig.py (hunk 3 above)
Comment (xmonader) [reply Jumpscale/tools/configmanager/DbConfig.py]: I believe it just loads the `TEMPLATE` variable in the clients and load it into self._template variable with 2 cases for the client: having child_class or not (Not sure what is child_class anyhow)

File: Jumpscale/data/nacl/NACL.py (hunk 1 above)
Comment (zaibon) [reply Jumpscale/data/nacl/NACL.py]: maybe not but this is cleaner, at least we only do work when it is required

File: Jumpscale/tools/configmanager/DbConfig.py (hunk 3 above)
Comment (zaibon) [reply Jumpscale/tools/configmanager/DbConfig.py]: so you didn't wrote this ? Can you blame the code you copied from and ping the author here ?
better clean this up while we are at it

Comment (xmonader) [reply Jumpscale/tools/configmanager/DbConfig.py]: @despiegk

File: Jumpscale/tools/configmanager/DbConfig.py [hunk 6]
```diff
@@ -107,6 +109,23 @@ def find_configs(client):
 
     return configs
 
+def template_from_object(obj):
+    module = None
```
Comment (zaibon) [inline Jumpscale/tools/configmanager/DbConfig.py]: add a doctring please
and you don't return any value in this function

**Pull request merged**